### ISSUE TITLE
Ver.2.0 "Failed to generate SSPI context" when IntegratedSecurity=true with Microsoft Account

### ISSUE THREAD

**Issue by akpapa** [dotnet/SqlClient#613]
Ver.2.0 **"Failed to generate SSPI context"** when IntegratedSecurity=true with Microsoft Account on Windows 10.
I tested SqlClient Ver.2.0 and 1.1.3 on two PCs, say PC1 and PC2.
- SQL Server 2019 Express: on PC1 Windows 10 build 20150.
- Client 1: PC1(the server),  **Microsoft account.** 
- Client 2: PC2, Windows 10 2004, **local account.**
- Both Clients: .Net Core 3.1,  IntegratedSecurity=true.
         Connection string: "Data Source=192.168.0.111;Initial Catalog=helloDb;Integrated Security=True"

Then I tried the following code from both Clients:
`await conn.OpenAsync().ConfigureAwait(false);`

Results:
- Ver. 2.0:
  (specifically from Preview Release **v2.0.0-preview4.20142.4**)
   PC1: **"Failed to generate SSPI context."**,
   PC2: no problem.
- Ver. 1.1.3
   Both PCs have no problem.

I read the release note of Ver. 2.0 but couldn't find any clue maybe because I am still learning SQL Server programming and not so familiar with it.
Any help would be appreciated.

**Comment by cheenamalhotra** [comment]
Hi @akpapa 

I tested on a machine logging in with Microsoft Account, I don't see any problem connecting to SQL 2019 Express.
Please note, if you have SQLEXPRESS installed, connection string would look like:

`Server=localhost\\SQLEXPRESS; Integrated Security=true;`

I also compared SSMS and SqlClient driver 2.0.0 both are resolving to same user account if you are connecting rightly:

![image](https://user-images.githubusercontent.com/13396919/85176905-c36b0800-b22f-11ea-8383-187bb2a61c93.png)

Let me know if you can share me more details to reproduce the problem.

**Comment by akpapa** [comment]
Thank you @cheenamalhotra for the test and kind response!
Sorry I omitted some details I thought irrelevant because local account can connect without issues.
Please refer to the details as below:

- The PC has SQL Server installed has a static IP address:192.168.0.111
- the instance has been named: SQLTEST.
- the port has been changed to 1234.
- connection string:
>  "Data Source=192.168.0.111,1234;Initial Catalog=helloDb;Integrated Security=True"
- unused IP addresses including 127.0.0.1 have been removed. 
![TCP_IP_Prop (Small)](https://user-images.githubusercontent.com/15343379/85187910-54002180-b2de-11ea-885c-576f4c9357e4.JPG)
### All tests are with logging in Microsoft account
**Tests with SqlClient v2.0.0-preview4.20142.4**
  *problem occurs since preview 4.
- PCNAME,1234\SQLTEST: works
- PCNAME\SQLTEST: works
- PCNAME,1234: works
- PCNAME: not work
  > The server was not found or was not accessible.
- localhost\SQLTEST: works
  (I don't know why it works, 127.0.0.1 has been removed from the IP address list.)


- 192.168.0.111,1234\SQLTEST: **not work**
  > Failed to generate SSPI context.
- 192.168.0.111\SQLTEST: **not work**
   >  The server was not found or was not accessible.
- 192.168.0.111,1234: **not work**
  > Failed to generate SSPI context.

**Tests with SSMS**
Same as with SqlClient.

Simply put, PCNAME works but IP address doesn't.
May be it is supposed to work this way as SSMS does the same.
The interesting thing is with SqlClient 1.1.3, IP address worked perfectly, so does other PC with local account in the LAN(same workgroup no domain) with 2.0, maybe that was not the right way and has been fixed in 2.0?
Thank you and enjoy your weekend!

**Comment by cheenamalhotra** [comment]
@akpapa 

I tested with IP address and port enabled on **TCP/IP** (I also disabled Shared memory and Named Pipes to be sure), and I could connect with below connection strings for both **v1.1.3** and **v2.0.0**:

```cs
Server=DESKTOP-9EBG74F,1234; Integrated Security=true;
Server=DESKTOP-9EBG74F,1234\SQLEXPRESS; Integrated Security=true;
Server=10.0.2.15,1234\SQLEXPRESS; Integrated Security=true;
Server=10.0.2.15,1234; Integrated Security=true;
Server=localhost,1234\SQLEXPRESS; Integrated Security=true;
Server=localhost,1234; Integrated Security=true;
```

It seems since you're able to connect to `PCNAME\SQLTEST` and `localhost\SQLTEST` they are defaulting to 1433 port as in my case they don't work in SSMS nor in application. I can only connect with port specified. Maybe something is mixed up on setup side at your end?

Do you have a use-case in your environment that is working with **SSMS** but not with Microsoft.Data.SqlClient v2.0.0 (on .NET Core/.NET Framework) if you connect specifically with TCP protocol and have disabled others?

**Comment by akpapa** [comment]
@cheenamalhotra , thanks for your prompt reply.

> I also disabled Shared memory and Named Pipes to be sure

 **Shared memory** is enabled,  **Named Pipes** is disabled.
I think they are the default settings.

I left settings untouched as possible.
Now I just changed port settings in "SQL Native Client 11.0 Configuration" and its 32bit setting to 1234 from1433, and I can still connect to PCNAME\InstanceName.
They are all settings I could find to change the port, even I don't know what they are for.

>Do you have a use-case in your environment that is working with SSMS but not with Microsoft.Data.SqlClient v2.0.0 (on .NET Core/.NET Framework) if you connect specifically with TCP protocol and have disabled others?

No, everything works with SSMS works with SqlClient 2.0, 
       everything does not work with SSMS does not work with SqlClient 2.0.
Instead, with SQLClient 1.1.3, I can connect to **192.168.0.111,1234** while SSMS cannot(SSPI context error).

**Comment by akpapa** [comment]
@cheenamalhotra 
Forgive me, I knew nothing about ***Shared Memory*** until I did some basic research.
After disabled ***Shared Memory***, now with only ***TCP/IP*** enabled, I can now connect to PCNAME with port explicitly specified, with or without instance name.
Still, connections by IP address get SSPI context error, both application v2.0 and SSMS.

**Comment by akpapa** [comment]
@cheenamalhotra 
Finally got the culprit and now I can connect to IP address with v2.0.
[This article](https://support.microsoft.com/en-us/help/811889/how-to-troubleshoot-the-cannot-generate-sspi-context-error-message) also helped a lot.
The hosts file modified by **docker** is responsible.
If I `ping PCNAME`, I got 192.168.0.111, but if I `ping -a 192.168.0.111`, I got **host.docker.internal**.
I guess it prevents the client driver from generating valid SPN?
I uninstalled docker a long time ago and never thought it left this little gift for me.


BTW, there are two facts made me never doubt if there is something wrong with my settings.
- SqlClient v1.1.3 works well.(don't know why tho)
- I read [some examples](https://social.msdn.microsoft.com/Forums/sqlserver/en-US/fe929f53-4663-4eb4-933e-66f9f2c1b331/network-access-to-sql-server-2016-with-microsoft-account?forum=sqlsecurity) which claim that SQL Server is not working well with Microsoft account.
   It seems to occur when connecting from another PC other than the server, maybe it's another issue.

I should have read some basics before asking.
Thank you very much for your time and kind help!
